internal/crossplane/service_mariadb_database.go

```diff
@@ -0,0 +1,178 @@
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	serviceMariadbDatabase = "mariadb-k8s-database"
+
+	planName   = "mariadb-user"
+	secretName = "%s-password"
+
+	// instanceSpecParamsPath is the path to an instance's parameters
+	instanceSpecParamsPath = "spec.parameters"
+
+	// instanceParamsParentReferenceName is the name of an instance's parent reference parameter
+	instanceParamsParentReferenceName = "parent_reference"
+	// instanceSpecParamsParentReferencePath is the path to an instance's parent reference parameter
+	instanceSpecParamsParentReferencePath = instanceSpecParamsPath + "." + instanceParamsParentReferenceName
+)
+
+// MariadbDatabaseServiceBinder defines a specific Mariadb service with enough data to retrieve connection credentials.
+type MariadbDatabaseServiceBinder struct {
+	instance  *Instance
+	resources []corev1.ObjectReference
+	cp        *Crossplane
+	logger    lager.Logger
+}
+
+// NewMariadbDatabaseServiceBinder instantiates a Mariadb service instance based on the given CompositeMariadbInstance.
+func NewMariadbDatabaseServiceBinder(c *Crossplane, instance *Instance, logger lager.Logger) *MariadbDatabaseServiceBinder {
+	return &MariadbDatabaseServiceBinder{
+		instance:  instance,
+		resources: instance.Composition.GetResourceReferences(),
+		cp:        c,
+		logger:    logger,
+	}
+}
+
+// Bind creates a MariaDB binding composite.
+func (msb MariadbDatabaseServiceBinder) Bind(ctx context.Context, bindingID string) (Credentials, error) {
+	parentRef, err := msb.retrieveDBInstance()
+	if err != nil {
+		return nil, err
+	}
+
+	pw, err := msb.createBinding(
+		ctx,
+		bindingID,
+		msb.instance.Labels.InstanceID,
+		parentRef,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// In order to directly return the credentials we need to get the IP/port for this instance.
+	secret, err := msb.cp.getCredentials(ctx, parentRef)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			err = ErrInstanceNotReady
+		}
+		return nil, err
+	}
+
+	endpoint, err := mapMariadbEndpoint(secret.Data)
+	if err != nil {
+		return nil, err
+	}
+
+	creds := createCredentials(endpoint, bindingID, pw, msb.instance.Composition.GetName())
+
+	return creds, nil
+}
+
+func (msb MariadbDatabaseServiceBinder) createBinding(ctx context.Context, bindingID, instanceID, parentReference string) (string, error) {
+	pw, err := password.Generate()
+	if err != nil {
+		return "", err
+	}
+
+	labels := map[string]string{
+		InstanceIDLabel: instanceID,
+		ParentIDLabel:   parentReference,
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf(secretName, bindingID),
+			Namespace: msb.cp.namespace,
+			Labels:    labels,
+		},
+		Data: map[string][]byte{
+			xrv1.ResourceCredentialsSecretPasswordKey: []byte(pw),
+		},
+	}
+	err = msb.cp.client.Create(ctx, secret)
+	if k8serrors.IsAlreadyExists(err) {
+		// FIXME(mw): this looks buggy
```

Comment:
This can only ever happen, if a binding with the same UID already exists, correct? I guess this will be handled by the API anyway, and we could just error-out here?

Reply:
yeah I wondered as well. Worth a try.